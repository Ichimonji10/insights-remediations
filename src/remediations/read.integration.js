'use strict';

const _ = require('lodash');
const { request, auth } = require('../test');

describe('remediations', function () {
    describe('list', function () {
        test('list remediations', async () => {
            const {body, text} = await request
            .get('/v1/remediations?pretty')
            .expect(200);

            body.should.have.property('remediations');
            body.remediations.should.not.be.empty();
            _.map(body.remediations, 'id').should.eql([
                '178cf0c8-35dd-42a3-96d5-7b50f9d211f6',
                'e809526c-56f5-4cd8-a809-93328436ea23',
                'cbc782e4-e8ae-4807-82ab-505387981d2e',
                '66eec356-dd06-4c72-a3b6-ef27d1508a02'
            ]);

            expect(text).toMatchSnapshot();
        });

        test('does not leak data outside of the account', async () => {
            const {body} = await request
            .get('/v1/remediations?username=99999')
            .expect(200);

            body.should.have.property('remediations');
            body.remediations.should.be.empty();
        });

        test('does not leak data outside of the account (2)', async () => {
            const {body} = await request
            .get('/v1/remediations?user_id=99999')
            .set(auth.emptyInternal)
            .expect(200);

            body.should.have.property('remediations');
            body.remediations.should.be.empty();
        });

        describe('sorting', function () {
            const [r1, r2, r3, r4] = [
                '178cf0c8-35dd-42a3-96d5-7b50f9d211f6',
                'e809526c-56f5-4cd8-a809-93328436ea23',
                'cbc782e4-e8ae-4807-82ab-505387981d2e',
                '66eec356-dd06-4c72-a3b6-ef27d1508a02'
            ];

            test('default', async () => {
                const {body} = await request
                .get('/v1/remediations?pretty')
                .expect(200);
                _.map(body.remediations, 'id').should.eql([r1, r2, r3, r4]);
            });

            function testSorting (column, asc, ...expected) {
                test(`${column} ${asc ? 'ASC' : 'DESC'}`, async () => {
                    const {body} = await request
                    .get(`/v1/remediations?pretty&sort=${asc ? '' : '-'}${column}`)
                    .expect(200);
                    _.map(body.remediations, 'id').should.eql(expected);
                });
            }

            testSorting('updated_at', true, r4, r3, r2, r1);
            testSorting('updated_at', false, r1, r2, r3, r4);
            testSorting('name', true, r4, r3, r1, r2);
            testSorting('name', false, r2, r1, r3, r4);
            testSorting('issue_count', true, r1, r2, r3, r4);
            testSorting('issue_count', false, r4, r3, r1, r2);
            testSorting('system_count', true, r1, r3, r2, r4);
            testSorting('system_count', false, r2, r4, r1, r3);
        });
    });

    describe('get', function () {
        test('get remediation', async () => {
            const {text} = await request
            .get('/v1/remediations/66eec356-dd06-4c72-a3b6-ef27d1508a02?pretty')
            .expect(200);

            expect(text).toMatchSnapshot();
        });

        test('get remediation (2)', async () => {
            const {body} = await request
            .get('/v1/remediations/e809526c-56f5-4cd8-a809-93328436ea23')
            .expect(200);

            body.should.eql({
                id: 'e809526c-56f5-4cd8-a809-93328436ea23',
                name: 'Unnamed remediation',
                needs_reboot: false,
                auto_reboot: false,
                created_by: 'tuser@redhat.com',
                created_at: '2018-12-04T08:19:36.641Z',
                updated_by: 'tuser@redhat.com',
                updated_at: '2018-12-04T08:19:36.641Z',
                issues: [{
                    id: 'advisor:network_bond_opts_config_issue|NETWORK_BONDING_OPTS_DOUBLE_QUOTES_ISSUE',
                    description: 'Bonding will not fail over to the backup link when bonding options are partially read',
                    resolution: {
                        id: 'fix',
                        description: 'Correct Bonding Config Items',
                        resolution_risk: 3,
                        needs_reboot: false
                    },
                    systems: [{
                        id: '1f12bdfc-8267-492d-a930-92f498fe65b9',
                        hostname: '1f12bdfc-8267-492d-a930-92f498fe65b9.example.com',
                        display_name: 'null'
                    }, {
                        id: 'fc94beb8-21ee-403d-99b1-949ef7adb762',
                        hostname: 'fc94beb8-21ee-403d-99b1-949ef7adb762',
                        display_name: 'null'
                    }]
                }]
            });
        });
    });
});
