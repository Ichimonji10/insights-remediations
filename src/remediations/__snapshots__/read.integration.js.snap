// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remediations get get remediation 1`] = `
"{
    \\"id\\": \\"66eec356-dd06-4c72-a3b6-ef27d1508a02\\",
    \\"name\\": \\"remediation 1\\",
    \\"needs_reboot\\": true,
    \\"auto_reboot\\": true,
    \\"created_by\\": {
        \\"username\\": \\"tuser@redhat.com\\",
        \\"first_name\\": \\"Test\\",
        \\"last_name\\": \\"User\\"
    },
    \\"created_at\\": \\"2018-10-04T08:19:36.641Z\\",
    \\"updated_by\\": {
        \\"username\\": \\"tuser@redhat.com\\",
        \\"first_name\\": \\"Test\\",
        \\"last_name\\": \\"User\\"
    },
    \\"updated_at\\": \\"2018-10-04T08:19:36.641Z\\",
    \\"issues\\": [
        {
            \\"id\\": \\"advisor:network_bond_opts_config_issue|NETWORK_BONDING_OPTS_DOUBLE_QUOTES_ISSUE\\",
            \\"description\\": \\"Bonding will not fail over to the backup link when bonding options are partially read\\",
            \\"resolution\\": {
                \\"id\\": \\"fix\\",
                \\"description\\": \\"Correct Bonding Config Items\\",
                \\"resolution_risk\\": 3,
                \\"needs_reboot\\": false
            },
            \\"resolutions_available\\": 1,
            \\"systems\\": [
                {
                    \\"id\\": \\"fc94beb8-21ee-403d-99b1-949ef7adb762\\",
                    \\"hostname\\": \\"null\\",
                    \\"display_name\\": \\"null\\"
                }
            ]
        },
        {
            \\"id\\": \\"compliance:xccdf_org.ssgproject.content_rule_sshd_disable_root_login\\",
            \\"description\\": \\"Disable SSH Root Login\\",
            \\"resolution\\": {
                \\"id\\": \\"fix\\",
                \\"description\\": \\"Fix\\",
                \\"resolution_risk\\": 1,
                \\"needs_reboot\\": false
            },
            \\"resolutions_available\\": 1,
            \\"systems\\": [
                {
                    \\"id\\": \\"fc94beb8-21ee-403d-99b1-949ef7adb762\\",
                    \\"hostname\\": \\"null\\",
                    \\"display_name\\": \\"null\\"
                }
            ]
        },
        {
            \\"id\\": \\"vulnerabilities:CVE-2017-17713\\",
            \\"description\\": \\"CVE-2017-17713\\",
            \\"resolution\\": {
                \\"id\\": \\"fix\\",
                \\"description\\": \\"Update packages (CVE-2017-17713)\\",
                \\"resolution_risk\\": -1,
                \\"needs_reboot\\": true
            },
            \\"resolutions_available\\": 1,
            \\"systems\\": [
                {
                    \\"id\\": \\"fc94beb8-21ee-403d-99b1-949ef7adb762\\",
                    \\"hostname\\": \\"null\\",
                    \\"display_name\\": \\"null\\"
                }
            ]
        },
        {
            \\"id\\": \\"vulnerabilities:CVE_2017_6074_kernel|KERNEL_CVE_2017_6074\\",
            \\"description\\": \\"Kernel vulnerable to local privilege escalation via DCCP module (CVE-2017-6074)\\",
            \\"resolution\\": {
                \\"id\\": \\"kernel_update\\",
                \\"description\\": \\"Update system to the latest kernel and reboot\\",
                \\"resolution_risk\\": 3,
                \\"needs_reboot\\": true
            },
            \\"resolutions_available\\": 3,
            \\"systems\\": [
                {
                    \\"id\\": \\"1f12bdfc-8267-492d-a930-92f498fe65b9\\",
                    \\"hostname\\": \\"1f12bdfc-8267-492d-a930-92f498fe65b9.example.com\\",
                    \\"display_name\\": \\"null\\"
                },
                {
                    \\"id\\": \\"fc94beb8-21ee-403d-99b1-949ef7adb762\\",
                    \\"hostname\\": \\"null\\",
                    \\"display_name\\": \\"null\\"
                }
            ]
        },
        {
            \\"id\\": \\"vulnerabilities:RHSA-2018:0502\\",
            \\"description\\": \\"RHSA-2018:0502\\",
            \\"resolution\\": {
                \\"id\\": \\"fix\\",
                \\"description\\": \\"Update packages (RHSA-2018:0502)\\",
                \\"resolution_risk\\": -1,
                \\"needs_reboot\\": true
            },
            \\"resolutions_available\\": 1,
            \\"systems\\": [
                {
                    \\"id\\": \\"fc94beb8-21ee-403d-99b1-949ef7adb762\\",
                    \\"hostname\\": \\"null\\",
                    \\"display_name\\": \\"null\\"
                }
            ]
        }
    ]
}"
`;

exports[`remediations list list remediations 1`] = `
"{
    \\"remediations\\": [
        {
            \\"id\\": \\"178cf0c8-35dd-42a3-96d5-7b50f9d211f6\\",
            \\"name\\": \\"Remediation with suppressed reboot\\",
            \\"created_by\\": {
                \\"username\\": \\"tuser@redhat.com\\",
                \\"first_name\\": \\"Test\\",
                \\"last_name\\": \\"User\\"
            },
            \\"created_at\\": \\"2018-12-05T08:19:36.641Z\\",
            \\"updated_by\\": {
                \\"username\\": \\"tuser@redhat.com\\",
                \\"first_name\\": \\"Test\\",
                \\"last_name\\": \\"User\\"
            },
            \\"updated_at\\": \\"2018-12-05T08:19:36.641Z\\",
            \\"needs_reboot\\": true,
            \\"system_count\\": 1,
            \\"issue_count\\": 1
        },
        {
            \\"id\\": \\"e809526c-56f5-4cd8-a809-93328436ea23\\",
            \\"name\\": \\"Unnamed remediation\\",
            \\"created_by\\": {
                \\"username\\": \\"tuser@redhat.com\\",
                \\"first_name\\": \\"Test\\",
                \\"last_name\\": \\"User\\"
            },
            \\"created_at\\": \\"2018-12-04T08:19:36.641Z\\",
            \\"updated_by\\": {
                \\"username\\": \\"tuser@redhat.com\\",
                \\"first_name\\": \\"Test\\",
                \\"last_name\\": \\"User\\"
            },
            \\"updated_at\\": \\"2018-12-04T08:19:36.641Z\\",
            \\"needs_reboot\\": false,
            \\"system_count\\": 2,
            \\"issue_count\\": 1
        },
        {
            \\"id\\": \\"cbc782e4-e8ae-4807-82ab-505387981d2e\\",
            \\"name\\": \\"remediation 2\\",
            \\"created_by\\": {
                \\"username\\": \\"tuser@redhat.com\\",
                \\"first_name\\": \\"Test\\",
                \\"last_name\\": \\"User\\"
            },
            \\"created_at\\": \\"2018-11-04T08:19:36.641Z\\",
            \\"updated_by\\": {
                \\"username\\": \\"tuser@redhat.com\\",
                \\"first_name\\": \\"Test\\",
                \\"last_name\\": \\"User\\"
            },
            \\"updated_at\\": \\"2018-11-04T08:19:36.641Z\\",
            \\"needs_reboot\\": true,
            \\"system_count\\": 1,
            \\"issue_count\\": 2
        },
        {
            \\"id\\": \\"66eec356-dd06-4c72-a3b6-ef27d1508a02\\",
            \\"name\\": \\"remediation 1\\",
            \\"created_by\\": {
                \\"username\\": \\"tuser@redhat.com\\",
                \\"first_name\\": \\"Test\\",
                \\"last_name\\": \\"User\\"
            },
            \\"created_at\\": \\"2018-10-04T08:19:36.641Z\\",
            \\"updated_by\\": {
                \\"username\\": \\"tuser@redhat.com\\",
                \\"first_name\\": \\"Test\\",
                \\"last_name\\": \\"User\\"
            },
            \\"updated_at\\": \\"2018-10-04T08:19:36.641Z\\",
            \\"needs_reboot\\": true,
            \\"system_count\\": 2,
            \\"issue_count\\": 5
        }
    ]
}"
`;

exports[`remediations missing listing of remediations does not blow up 1`] = `
Object {
  "remediations": Array [
    Object {
      "created_at": "2018-11-04T08:19:36.641Z",
      "created_by": Object {
        "first_name": "Test",
        "last_name": "Single",
        "username": "testReadSingleUser",
      },
      "id": "82aeb63f-fc25-4eef-9333-4fa7e10f7217",
      "issue_count": 1,
      "name": "missing system 1",
      "needs_reboot": true,
      "system_count": 2,
      "updated_at": "2018-11-04T08:19:36.641Z",
      "updated_by": Object {
        "first_name": "Test",
        "last_name": "Single",
        "username": "testReadSingleUser",
      },
    },
    Object {
      "created_at": "2018-11-04T07:19:36.641Z",
      "created_by": Object {
        "first_name": "Test",
        "last_name": "Single",
        "username": "testReadSingleUser",
      },
      "id": "27e36e14-e1c2-4b5a-9382-ec80ca9a6c1a",
      "issue_count": 2,
      "name": "missing system 2",
      "needs_reboot": true,
      "system_count": 2,
      "updated_at": "2018-11-04T07:19:36.641Z",
      "updated_by": Object {
        "first_name": "Test",
        "last_name": "Single",
        "username": "testReadSingleUser",
      },
    },
    Object {
      "created_at": "2018-11-04T06:19:36.641Z",
      "created_by": Object {
        "first_name": "Test",
        "last_name": "Single",
        "username": "testReadSingleUser",
      },
      "id": "ea5b1507-4cd3-4c87-aa5a-6c755d32a7bd",
      "issue_count": 1,
      "name": "unknown resolution",
      "needs_reboot": true,
      "system_count": 1,
      "updated_at": "2018-11-04T06:19:36.641Z",
      "updated_by": Object {
        "first_name": "Test",
        "last_name": "Single",
        "username": "testReadSingleUser",
      },
    },
    Object {
      "created_at": "2018-11-04T05:19:36.641Z",
      "created_by": Object {
        "first_name": "Test",
        "last_name": "Single",
        "username": "testReadSingleUser",
      },
      "id": "62c95092-ac83-4025-a676-362a67e68579",
      "issue_count": 0,
      "name": "unknown issues",
      "needs_reboot": false,
      "system_count": 0,
      "updated_at": "2018-11-04T05:19:36.641Z",
      "updated_by": Object {
        "first_name": "Test",
        "last_name": "Single",
        "username": "testReadSingleUser",
      },
    },
  ],
}
`;
